File: R/mle.mglasso.R
mle.mglasso <- function(object, ..., maxit_em = 1.0e+03, thr_em = 1.0e-4, maxit_bcd = 1.0e+4, thr_bcd = 1.0e-4, trace = 0L) {
    this.call <- this.call <- match.call()
    # checking 'maxi_emt'
    if(!is.vector(maxit_em)) stop(sQuote("maxit_em"), " is not a vector of length ", sQuote(1))
    if(length(maxit_em) != 1) stop(sQuote("maxit_em"), " is not a vector of length ", sQuote(1))
    if(abs(as.integer(maxit_em) - maxit_em) > 0) stop(sQuote("maxit_em"), " is not an object of type ", dQuote("integer"))
    if(maxit_em <= 0) stop(sQuote("maxit_em"), " is not a positive integer")
    # checking 'thr_em'
    if(!is.vector(thr_em)) stop(sQuote("thr_em"), " is not a vector of length ", sQuote(1))
    if(length(thr_em) != 1) stop(sQuote("thr_em"), " is not a vector of length ", sQuote(1))
    if(thr_em < 0 ) stop(sQuote("thr_em"), " is not a vector of positive values")
    # checking 'maxit_bcd'
    if(!is.vector(maxit_bcd)) stop(sQuote("maxit_bcd"), " is not a vector of length ", sQuote(1))
    if(length(maxit_bcd) != 1) stop(sQuote("maxit_bcd"), " is not a vector of length ", sQuote(1))
    if(abs(as.integer(maxit_bcd) - maxit_bcd) > 0) stop(sQuote("maxit_bcd"), " is not an object of type ", dQuote("integer"))
    if(maxit_bcd <= 0) stop(sQuote("maxit_bcd"), " is not a positive integer")
    # checking 'thr_bcd'
    if(!is.vector(thr_bcd)) stop(sQuote("thr_bcd"), " is not a vector of length ", sQuote(1))
    if(length(thr_bcd) != 1) stop(sQuote("thr_bcd"), " is not a vector of length ", sQuote(1))
    if(thr_bcd < 0 ) stop(sQuote("thr_bcd"), " is not a vector of positive values")
    # checking 'trace'
    if(!is.vector(trace)) stop(sQuote("trace"), " is not a vector of length ", sQuote(1))
    if(length(trace) != 1) stop(sQuote("trace"), " is not a vector of length ", sQuote(1))
    if(abs(as.integer(trace) - trace) > 0) stop(sQuote("trace"), " is not an object of type ", dQuote("integer"))
    if(!is.element(trace, c(0L, 1L, 2L))) stop(sQuote("trace"), " can be equal to '0', '1' or '2'")
    # extracting elements from 'object'
    X <- object$X
    R <- is.na(X)
    X[R] <- .Machine$double.xmax
    X <- datacggm(X, up = .Machine$double.xmax)
    R <- X$R[-1L, -1L]
    startmis <- X$startmis
    X <- X$X
    n <- dim(X)[1L]
    p <- dim(X)[2L]
    nrho <- object$nrho
    Xipt <- object$Xipt
    S <- object$S
    mu <- object$mu
    Sgm <- object$Sgm
    Tht <- object$Tht
    R2 <- object$R2
    nit <- matrix(0L, nrho, 2, dimnames = list(NULL, c("EM", "nit")))
    conv <- object$conv
    subrout <- object$subrout
    # storage.mode setting
    storage.mode(n) <- "integer"
    storage.mode(p) <- "integer"
    storage.mode(X) <- "double"
    storage.mode(R) <- "integer"
    storage.mode(startmis) <- "integer"
    storage.mode(nrho) <- "integer"
    storage.mode(maxit_em) <- "integer"
    storage.mode(thr_em) <- "double"
    storage.mode(maxit_bcd) <- "integer"
    storage.mode(thr_bcd) <- "double"
    storage.mode(Xipt) <- "double"
    storage.mode(S) <- "double"
    storage.mode(mu) <- "double"
    storage.mode(Sgm) <- "double"
    storage.mode(Tht) <- "double"
    R2 <- double(nrho)
    storage.mode(nit) <- "integer"
    conv <- integer(1L)
    subrout <- integer(1L)
    storage.mode(trace) <- "integer"
    out <- .Fortran(C_mlemglasso, n = n, p = p, X = X, R = R, startmis = startmis,
                    nrho = nrho, maxit_em = maxit_em, thr_em = thr_em,
                    maxit_bcd = maxit_bcd, thr_bcd = thr_bcd, Xipt = Xipt, S = S, mu = mu,
                    Sgm = Sgm, Tht = Tht, R2 = R2, nit = nit, conv = conv, subrout = subrout,
                    trace = trace)
    # removing unused elements
    out <- out[-c(1L, 2L, 4L, 5L)]
    # adding elements form 'object'
    out$X[R[, -1L] == 1L] <- NA
    out$Adj <- object$Adj
    out$rho <- object$rho
    out$Ck <- object$Ck
    out$pk <- object$pk
    out$ncomp <- object$ncomp
    out$df <- object$df
    # checking convergence
    if(out$conv != 0L) {
        nrho <- out$nrho
        out$subrout <- switch(as.character(out$subrout),
                        "0" = "Ok",
                        "1" = "impute",
                        "2" = "glasso",
                        "3" = "mlemglasso")
        out$conv <- switch(as.character(out$conv),
                        "-1" = "memory allocation error",
                        "1" = "maximum number of iterations has been exceeded",
                        "2" = "error in computing the coditional moments",
                        "3" = "matrix inversion failed")
        if(nrho == 0) stop("mle.mglasso does not congerce: ", sQuote(out$conv), "\n  Error in subroutine ", sQuote(out$subrout))
        msg <- paste("mle.mglasso does not congerce at nrho =", nrho + 1L, " with error code", sQuote(out$conv), "\n  Error in subroutine ", sQuote(out$subrout))
        msg <- paste(msg, ifelse(nrho == 1, "\n  The first solution is reported", paste("\n  The first", nrho, "solutions are reported")))
        warning(msg)
        out$rho <- out$rho[1:nrho]
        out$Xipt <- out$Xipt[, , 1:nrho, drop = FALSE]
        out$S <- out$S[, , 1:nrho, drop = FALSE]
        out$mu <- out$mu[, 1:nrho, drop = FALSE]
        out$Sgm <- out$Sgm[, , 1:nrho, drop = FALSE]
        out$Tht <- out$Tht[, , 1:nrho, drop = FALSE]
        out$Adj <- out$Adj[, , 1:nrho, drop = FALSE]
        out$df <- out$df[1:nrho]
        out$R2 <- out$R2[1:nrho]
        out$ncomp <- out$ncomp[1:nrho]
        out$Ck <- out$Ck[, 1:nrho, drop = FALSE]
        out$pk <- out$pk[, 1:nrho, drop = FALSE]
        out$nit <- out$nit[1:nrho, , drop = TRUE]
    }
    out <- c(call = this.call, out)
    class(out) <- c("mggm", "mglasso", "glasso")
    out
}

File: src/mlemglasso.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Author: Luigi Augugliaro
! e-mail: luigi.augugliaro@unipa.it
! webpage: http://dssm.unipa.it/augugliaro
!
! version: 1.0.0
! Data: September 17, 2018
!
subroutine mlemglasso(n,p,X,R,startmis,nrho,maxit_em,thr_em,maxit_bcd,thr_bcd,Xipt,S,mu,Sgm,Tht,R2,nit,conv,subrout,trace)
implicit none
integer :: n,p,R(n,0:p),startmis,nrho,maxit_em,maxit_bcd,nit(nrho,2),conv,subrout,trace
double precision :: X(n,p),thr_em,thr_bcd
double precision :: Xipt(n,p,nrho),S(p,p,nrho),mu(p,nrho),Sgm(p,p,nrho),Tht(p,p,nrho),R2(nrho)
! internal variables
integer :: i,j,k,ii,nnit,ncomp,Ck(p),pk(p)
double precision, parameter :: big = huge(1.d0)
double precision :: T1o(p),T2o(p,p),T1(p),T2(p,p),mu_o(p),Sgm_o(p,p),Tht_o(p,p),S_k(p,p),mu_k(p),Sgm_k(p,p),Tht_k(p,p)
double precision :: Xipt_k(n,p),rho_gl(p,p),dmu,dTht,R2num,R2den
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! computing the observed statistics T1o and T2o !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
T1o = 0.d0
do j = 1, p
    do i = 1, n
        if(R(i, j).eq.0) then
            T1o(j) = T1o(j) + X(i, j)
        end if
    end do
end do
T2o = 0.d0
do i = 1, p
    do j = i, p
        do k = 1, n
            if((R(k, i).eq.0).and.(R(k, j).eq.0)) then
                T2o(i, j) = T2o(i, j) + X(k, i) * X(k, j)
            end if
        end do
        T2o(j, i) = T2o(i, j)
    end do
end do
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Starting optimization                  !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Xipt_k = X
mu_k = mu(:, 1)
Sgm_k = Sgm(:, :, 1)
Tht_k = Tht(:, :, 1)
do k = 1, nrho
    if(trace.eq.2) call mlemglasso_trace2_1(k)
    rho_gl = 0.d0
    do j = 1, p
        do i = j + 1, p
            if(abs(Tht(i, j, k)).eq.0.d0) then
                rho_gl(i, j) = big
                rho_gl(j, i) = big
            end if
        end do
    end do
    !!!!!!!!!!!!!!!!!!!!!!!!!
    ! starting EM algorithm !
    !!!!!!!!!!!!!!!!!!!!!!!!!
    mu_o = mu_k
    Sgm_o = Sgm_k
    Tht_o = Tht_k
    do ii = 1, maxit_em
        if(trace.eq.2) call cglasso_trace2_2(ii)
        call impute(startmis, n, p, X, R, T1o, T2o, mu_o, Tht_o, Xipt_k, T1, T2, conv)
        if(conv.ne.0) then
            subrout = 1
            exit
        end if
        mu_k = T1 / n
        do j = 1, p
            S_k(:, j) = T2(:, j) / n - mu_k * mu_k(j)
        end do
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        ! inizializing the matrix Sgm_k !
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        Sgm_k = S_k
        do j = 1, p
            do i = j + 1, p
                if(abs(Tht_o(i, j)).gt.0.d0) then
                    Sgm_k(i, j) = Sgm_k(i, j) + rho_gl(i, j) * sign(1.d0, Tht_o(i, j))
                    Sgm_k(j, i) = Sgm_k(i, j)
                end if
            end do
        end do
        if(trace.eq.2) call cglasso_trace2_3()
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        ! solving the glasso problem  !
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        nnit = 0
        ncomp = 0
        Ck = 0
        pk = 0
        if(trace.eq.2) call mlemglasso_trace2_4()
        call glassosub(p, S_k, 0, rho_gl, maxit_bcd, thr_bcd, Sgm_k, Tht_k, ncomp, Ck, pk, nnit, conv, trace)
        if(conv.ne.0) then
            subrout = 2
            exit
        end if
        if(trace.eq.2) call cglasso_trace2_5()
        nit(k, 1) = ii
        nit(k, 2) = nit(k, 2) + nnit
        Sgm_k = (Sgm_k + transpose(Sgm_k)) / 2.d0
        Tht_k = (Tht_k + transpose(Tht_k)) / 2.d0
        dmu = maxval(abs(mu_o - mu_k))
        dTht = 0.d0
        do j = 1, p
            do i = 1, j
                dTht = max(dTht, abs(Tht_o(i, j) - Tht_k(i, j)))
            end do
        end do
        if(trace.eq.2) call cglasso_trace2_6(thr_em, dmu, dTht)
        if(max(dmu, dTht).le.thr_em) then
            exit
        else
            mu_o = mu_k
            Tht_o = Tht_k
        end if
    end do
    if(ii.ge.maxit_em) then
        conv = 1
        subrout = 3
    end if
    if(conv.eq.0) then
        if(trace.eq.1) call mlemglasso_trace1(k, nit(k, 1), nit(k, 2))
        Xipt(:, :, k) = Xipt_k
        S(:, :, k) = S_k
        mu(:, k) = mu_k
        Sgm(:, :, k) = Sgm_k
        Tht(:, :, k) = Tht_k
        R2num = 0.d0
        R2den = 0.d0
        do j = 1, p
            R2num = R2num + (S_k(j, j) - Sgm_k(j, j))**2
            do i = j + 1, p
                R2num = R2num + (S_k(i, j) - Sgm_k(i, j))**2
                R2den = R2den + S_k(i, j)**2
            end do
        end do
        R2(k) = 1.d0 - R2num / R2den
    else
        nrho = k - 1
        exit
    end if
end do
end subroutine mlemglasso
